python/cudf/cudf/pandas/_wrappers/pandas.py

```diff
@@ -1893,6 +1869,106 @@ def initial_setup():
     cudf.set_option("mode.pandas_compatible", True)
 
 
+def _reduce_obj(obj):
+    from cudf.pandas.module_accelerator import disable_module_accelerator
+
+    with disable_module_accelerator():
+        pickled_args = pickle.dumps(obj.__reduce__())
+
+    return _unpickle_obj, (pickled_args,)
+
+
+def _unpickle_obj(pickled_args):
+    from cudf.pandas.module_accelerator import disable_module_accelerator
+
+    with disable_module_accelerator():
+        unpickler, args = pickle.loads(pickled_args)
+    obj = unpickler(*args)
+    return obj
+
+
+def _generic_reduce_obj(obj, unpickle_func):
+    from cudf.pandas.module_accelerator import disable_module_accelerator
+
+    with disable_module_accelerator():
+        pickled_args = pickle.dumps(obj.__reduce__())
+
+    return unpickle_func, (pickled_args,)
+
+
+def _frame_unpickle_obj(pickled_args):
+    from cudf.pandas.module_accelerator import disable_module_accelerator
+
+    with disable_module_accelerator():
+        unpickled_intermediate = pickle.loads(pickled_args)
+        reconstructor_func = unpickled_intermediate[0]
+        obj = reconstructor_func(*unpickled_intermediate[1])
+        obj.__setstate__(unpickled_intermediate[2])
+    return obj
+
+
+def _index_unpickle_obj(pickled_args):
+    from cudf.pandas.module_accelerator import disable_module_accelerator
+
+    with disable_module_accelerator():
+        unpickled_intermediate = pickle.loads(pickled_args)
+        reconstructor_func = unpickled_intermediate[0]
+        obj = reconstructor_func(*unpickled_intermediate[1])
+
+    return obj
+
+
+def _reduce_offset_obj(obj):
+    from cudf.pandas.module_accelerator import disable_module_accelerator
+
+    with disable_module_accelerator():
+        pickled_args = pickle.dumps(obj.__getstate__())
+
+    return _unpickle_offset_obj, (pickled_args,)
+
+
+def _unpickle_offset_obj(pickled_args):
+    from cudf.pandas.module_accelerator import disable_module_accelerator
+
+    with disable_module_accelerator():
+        data = pickle.loads(pickled_args)
+        data.pop("_offset")
+        data.pop("_use_relativedelta")
+    obj = pd._libs.tslibs.offsets.DateOffset(**data)
+    return obj
+
+
 copyreg.dispatch_table[pd.Timestamp] = _reduce_obj
 # same reducer/unpickler can be used for Timedelta:
 copyreg.dispatch_table[pd.Timedelta] = _reduce_obj
+
+# TODO: Need to find a way to unpickle cross-version(old) pickled objects.
+# Register custom reducer/unpickler functions for pandas objects
+# so that they can be pickled/unpickled correctly:
+copyreg.dispatch_table[pd.Series] = lambda obj: _generic_reduce_obj(
```

Comment:
Perhaps I'm misunderstanding the code here, but aren't we registering these dispatches for true pandas objects? Don't we want them registered for cudf.pandas proxy types?  I assume that what you've done is correct based on the preexisting dispatches on lines 1941 and 1943, but I don't follow why.

Reply:
We have `_FinalProxy.__reduce__` and `_FinalProxy.__setstate__` defined for proxy objects. Ideally, we want these methods to be invoked by pickle.dump and pickle.load. The issue is we are relying on pandas serialization logic in `_FinalProxy.__reduce__` to serialize. Which writes true pandas objects into pickle files. Now when we unpickle that file, we are trying to load a true pandas object, for which we will need to register custom unpickling methods so that we can convert them to `cudf.pandas` proxy objects.